### PR TITLE
ci(typecheck): typecheck the plugin packages in CI (#987)

### LINKED ISSUE #987: Typecheck the plugin packages in CI — rollup-only builds let type errors reach master
## Background

While resolving #926 it turned out that **the plugin packages are never typechecked**, so type errors reach `master` unnoticed.

Every plugin package's build script is transpile-only:

```json
"build": "rollup -c && npm run postbuild"
```

`rollup -c` transpiles TypeScript without checking it, and CI runs only ESLint, Prettier, and Vitest. Vitest's esbuild path is likewise transpile-only. So a green `pnpm build` plus a green CI run says nothing about the type correctness of `iracing-plugin-stream-deck`, `iracing-plugin-mirabox`, or `iracing-plugin-ulanzi` — the three packages that wire everything together and where cross-package integration errors actually surface.

This is not hypothetical. Running `tsc --noEmit` by hand during #926 found:

- Two `TS2300` duplicate-binding errors in `iracing-plugin-stream-deck/src/shared/index.ts`, introduced by a merge and caught only because CodeRabbit read the diff.
- A `TS2322` in `plugin.ts` that had been **sitting on `master` since #930**: `initWindowFocus(…, () => native.focusIRacingWindow())`, where the native wrapper returned bare `number` but the deck-core delegate type demanded `FocusResult`. Nothing flagged it for two PRs.

Library packages (`deck-core`, `iracing-sdk`, `icons`, …) do run `tsc` as their build, so this gap is specific to the three plugin packages plus anything else whose build is rollup-only.

## Proposed change

1. Add a `typecheck` script (`tsc --noEmit -p tsconfig.json`) to every package whose `build` does not already run `tsc` — the three plugins at minimum; audit the rest.
2. Add a root `typecheck` script wired through turbo so `pnpm typecheck` covers the workspace.
3. Add a **Typecheck** job to the CI workflow alongside ESLint / Prettier / Run tests.
4. Consider making `pnpm build` (or the `precommit` script) depend on it, so the gap can't reappear locally.

**Blocked by #988** — `tsc --noEmit` currently reports pre-existing `TS2823` errors ("Import attributes are only supported when the '--module' option is set to …") in `iracing-actions` and `plugin.ts`. Those must be resolved first, or the new job fails on day one.

## Affected artifacts beyond the code

- **All plugin packages** — a `typecheck` script in each `package.json`.
- **CI** — a new job in the workflow, and `turbo.json` task wiring.
- **Rules and guidance** — `.claude/rules/build-and-commit.md` (pre-commit checks) and `.claude/rules/testing.md` should state that `pnpm build` alone does not typecheck the plugins, and what does.

### PR DESCRIPTION
**Pull request by niklam**
## Related Issue

Fixes #987

## What changed?

The three plugins and `pi-components` build through rollup, which transpiles without checking, and CI ran only ESLint, Prettier and Vitest — whose esbuild path is also transpile-only. So a green build and a green CI run said nothing about the type correctness of the four packages where cross-package integration actually surfaces.

### The defect that ships today, reproduced on demand

`@rollup/plugin-typescript` routes every diagnostic through `context.warn` unless `noEmitOnError` is set — **errors included** — so the build emits and exits 0. Planting the exact defect from #987 (an identifier never imported) and running it both ways:

```text
WITH noEmitOnError      build exit=1
                        RollupError: TS2304: Cannot find name 'MISSING_IDENTIFIER_PROBE'
                        (names the file and line: src/plugin.ts:1134)

WITHOUT it (today)      build exit=0
                        TS2304 appears once — as a warning
                        the identifier is in the emitted plugin.js: 1 occurrence
```

The second half is the argument. **This is not "we added a gate"; it is "here is a green build shipping a startup `ReferenceError`, on demand."** Verified separately for `iracing-plugin-stream-deck` and for `pi-components`, whose `browser/pi-components.js` — the PI framework every plugin loads — was equally able to ship an undefined identifier. Both restored and rebuilt afterwards, probes verified gone.

The counterfactual is what isolates the cause: it shows the flag is load-bearing rather than merely present alongside a passing build.

### The gate

- **`typecheck` script in 20 packages**, and a root `pnpm typecheck` over them via turbo.
- **The turbo task declares `dependsOn: ["^build"]`**, because packages resolve each other's *types* through emitted `dist/`. Putting the constraint there rather than in the workflow means CI runs the same one command a developer does, so a red run reproduces locally with `pnpm typecheck` — no folklore about building first.
- **`noEmitOnError` on all four rollup configs.** `pi-components` gets a small `tsBundle` helper so the option and its reasoning live in one place rather than four.
- **A Typecheck workflow**, triggering on `pull_request` **and** on pushes to `master` and `release/*`. A PR check only proves something about the branch merged against the master that existed when it ran (#1070); the defect class this gate targets is exactly what two PRs merging in succession produce, so a gate that never sees the merged result would miss its highest-value case. Release branches take feature PRs of their own and reach master by back-merge, so the same argument covers them. This does **not** create a convention for who fixes a red master — that belongs to #1070.

### Every exclusion is a named entry that expires

Two guards, each asserting one invariant, and **both fail in the reverse direction** — which is what separates this from a suppression list:

| Guard | Catches | Reverse direction |
|---|---|---|
| Every package with TypeScript is in the gate | a package the gate never runs on | an excluded package gaining a `typecheck` script fails as *outlived* |
| A package's typecheck covers its own test files | a package it runs on while checking none of its tests | an excused package that stops excluding fails as *outlived* |

`turbo run typecheck` skips a package with no script **silently** — no error, exit 0 — so coverage could shrink to nothing while the gate still reported success. That is this issue's own defect one level up, which is why it is guarded mechanically rather than by care.

Every entry carries the size of what it hides and its maturity date: **#1077** (website — Astro, `astro check` needs an uninstalled dependency) and **#1078** (`iracing-actions` has no tsconfig; two adapters exclude their own tests, hiding 34 and 29 errors). `.claude/rules/testing.md` names the same two, so the guard and the docs cannot drift about what is excluded.

The second invariant resolves each tsconfig through TypeScript's own `parseJsonConfigFileContent` rather than pattern-matching its `exclude` array, so it reports what is genuinely in the program under the compiler's own semantics.

### pi-components: eight errors, three causes

Nothing had ever typechecked it. The eight errors were not eight defects:

- **`allowJs`** on its typecheck-only project clears five — the four `TS7016` from untyped `src/build/*.mjs`, and the `TS7006` that followed from one of them. Reading the real sources beats hand-written `.d.mts` declarations, which are a second description of something that already describes itself. **This cannot affect any build output**: the four bundles use the sub-configs, never this `noEmit: true` project.
- **Two `TS2307`** — the package imported `@iracedeck/deck-core` from two test files while declaring it nowhere, resolving only through workspace hoisting. Declared as a devDependency; no cycle. Test-only, so not shipped.
- **One `TS2352`** — fixed at the cause. A bare `vi.fn()` has no call signature, so `mock.calls[0]` was `[]` and the test cast it to a five-tuple. Giving the mock sdpi's real connect signature makes the cast **disappear** rather than compile. Silencing a type error in the PR that exists to stop type errors being silenced would be self-defeating.

### A rule file was asserting the opposite of this PR

`.claude/rules/build-and-commit.md`'s **Build verification** section told every contributor that "the build may succeed (exit code 0) while still emitting TypeScript warnings" and to grep the output for `TS[0-9]+:`. With `noEmitOnError` those are fatal.

Worth stating plainly: **I corrected a different, smaller false claim in the same file first, and left this one — 93 lines above it — until code review found it.** So the file contradicted itself, in the pre-commit checklist everyone follows, about this exact defect class. That is the strongest available evidence that the gate needed to exist, and it is not evidence I produced.

`precommit` also ran `build` without `typecheck` while the docs had just started listing it as step 3; fixed.

## How to test

```bash
pnpm install && pnpm build
pnpm typecheck              # 37 tasks — 20 typechecks plus the builds turbo pulls in
pnpm test                   # 327 files / 9325 tests
pnpm lint && pnpm format
```

To see the gate work rather than pass, plant a type error in any plugin's `plugin.ts` and confirm both `pnpm typecheck` and `pnpm build` go red naming the file — then remove it.

Green on `f493d92f`: build 22/22 with `0 cached`, typecheck 37/37, lint 0, format clean, **327 files / 9325 tests** — the 9278 master baseline plus this PR's 47 guard cases.

Rebased onto `f493d92f` after #960 merged, and the counterfactual above was **re-run on the rebased head** rather than carried forward — a control measured against a base nobody merges from is a claim about a tree that will not exist. Identical result. In CI the new job took **1m28s**, finishing before the existing test job's 2m9s, so it adds no wall clock to PR feedback.

## Notes for the reviewer

- The five commits are kept separate so the response to code review is visible as such; squash collapses them anyway.
- Code review raised **seven** findings and all seven held. Three mattered: the docs self-contradiction above; the coverage guard being keyed on `existsSync(tsconfig.json)`, which let three packages escape it entirely rather than appear as exclusions; and presence-is-not-coverage. Fixed in `b5609f5c` and `624f7434`.
- No changelog entry: internal CI and tooling, no user-visible effect.

## Checklist

- [x] Linked to an approved issue
- [x] New code has unit tests — two guards, 47 cases, each controlled by being made to fail on purpose in both directions before being trusted green
- [x] Changes registered in all applicable plugins (Stream Deck, Mirabox) — and Ulanzi; `noEmitOnError` on all four rollup configs, `typecheck` in all 20 checkable packages
- [x] No unrelated changes included

### CODE REVIEW

File: scripts/typecheck-script-coverage.test.mjs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+]);
+
+// Resolve a tsconfig the way tsc does, rather than pattern-matching its `exclude`
+// array: `parseJsonConfigFileContent` applies the same include/exclude/files
+// semantics the compiler will, so this reports what is genuinely in the program.
+function filesInProgram(packageName) {
+  const dir = join(repoRoot, "packages", packageName);
+  const configPath = join(dir, "tsconfig.json");
+  const { config } = ts.parseConfigFileTextToJson(configPath, readFileSync(configPath, "utf-8"));
+  return new Set(ts.parseJsonConfigFileContent(config ?? {}, ts.sys, dir).fileNames.map(toPosix));
```
[the remark this answers is not in the document]
Comment (niklam) [reply scripts/typecheck-script-coverage.test.mjs]: Applied in `384f979e` — and this was a real silent pass, not a theoretical one.

I reproduced it before accepting: a tsconfig whose `extends` points at a nonexistent file gave **`47 passed`**. A genuinely broken config, reported fully covered, by the guard whose entire purpose is noticing coverage that is not there. Exactly the mechanism you described — the discarded error, `config ?? {}`, and `parseJsonConfigFileContent` then defaulting to "every `.ts` under basePath" so every test file looked present.

Malformed JSON did fail, but on an internal `Debug Failure` about path separators rather than anything actionable, so that is fixed too by normalising the path before handing it to TypeScript.

Both paths now throw naming the file and the diagnostic:

```
packages/logger/tsconfig.json did not resolve cleanly: Cannot read file .../does-not-exist.json.
packages/logger/tsconfig.json could not be parsed: '{' expected.
```

Verified by re-running both cases after the fix.

File: .github/workflows/ci-typecheck.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  pull_request:
+  push:
+    branches: [master, 'release/*']
+
+jobs:
+  typecheck:
+    name: Typecheck
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v7
```
[the remark this answers is not in the document]
Comment (niklam) [reply .github/workflows/ci-typecheck.yml:30]: Applied in `384f979e`: job-level `permissions: contents: read`, and `persist-credentials: false` on the checkout. The job reads the repository and runs a compiler — it needs no token scope beyond read and no credential left in `.git/config` for later steps to reach.

One deliberate note in case it looks inconsistent: the three sibling CI workflows (`ci-test`, `ci-lint`, `ci-format`) set neither, so this workflow does not match them. That is intentional rather than an oversight — copying a weaker default forward is not consistency worth having, and three other workflows in the repo (`pr-labeler`, `release`, `release-pack`) already declare `permissions`, so this is not foreign to the codebase. Bringing the other three into line is out of scope here but worth doing.

**Comment by niklam** [comment]
All three review findings applied in `384f979e`, each reproduced before being accepted. Replies on the two inline threads; the nitpick is answered here.

**`.tsx` in both discovery checks — applied.** Worth being explicit that this one is speculative: the repo currently contains **zero** `.tsx` files, so nothing is broken today. I took it anyway because the guard's whole subject is coverage that appears complete and is not, and a package whose only sources were `.tsx` would be invisible to it — the same silent escape, one file extension along. `.d.tsx` is excluded alongside `.d.ts`.

**Also, since the base moved under this PR.** #960 merged while I was pushing, so the branch is rebased onto `f493d92f` and everything was re-measured rather than carried forward — including the counterfactual, which is the load-bearing claim here. A control measured against a base nobody merges from is a claim about a tree that will not exist. Identical result on the rebased head: with `noEmitOnError` the build exits 1 naming the file and line; without it, exit 0 with the undefined identifier in the emitted `plugin.js`.

Green on `f493d92f`: typecheck 37/37, lint 0, format clean, **327 files / 9325 tests** (the 9278 master baseline plus this PR's 47 guard cases).

One number I chased rather than accepted: I predicted 9322 and measured 9325. The gap is that #960 rewrote `camera-controls.test.ts`, which uses `it.each` — my estimate counted new `it(` blocks but not new rows added to existing tables.

And the new job's first CI run passed in **1m28s**, finishing before the existing test job's 2m9s. Since the jobs run in parallel, the added wall clock to PR feedback is effectively zero — which was the argument for a separate job rather than folding typecheck into the test job, now measured rather than assumed.
